src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
+        above <vector> [by <scalar/vector>]
+
+    If the :grammar:`by <scalar/vector>` is omitted, zero is used.
+    """
+    return directionalSpecHelper('Above', pos, dist, 'height', lambda dist: (0, 0, dist),
+                          lambda self, dims, tol, dx, dy, dz: Vector(dx, dy, self.height / 2 + dz + dims[2]/2 + tol))
+
+def Below(pos, dist=0):
+    """The :specifier:`below {X} by {Y}` polymorphic specifier.
+
+    Specifies :prop`position`, depending on :prop:`height`.
+
+    Allowed forms::
+
+        below <oriented point> [by <scalar/vector>]
+        below <vector> [by <scalar/vector>]
+
+    If the 'by <scalar/vector>' is omitted, zero is used.
+    """
+    return directionalSpecHelper('Below', pos, dist, 'height', lambda dist: (0, 0, dist),
+                          lambda self, dims, tol, dx, dy, dz: Vector(dx, dy, -self.height / 2 - dz - dims[2]/2 - tol))
+
+def directionalSpecHelper(syntax, pos, dist, axis, toComponents, makeOffset):
+    prop = {'position': 1}
+    if canCoerce(dist, builtins.float):
+        dx, dy, dz = toComponents(coerce(dist, builtins.float))
+    elif canCoerce(dist, Vector):
+        dx, dy, dz = coerce(dist, Vector)
+    else:
+        raise TypeError(f'"{syntax} X by D" with D not a number or vector')
+
+    @distributionFunction
+    def makeContactTol(dist, ct):
+        if dist == 0:
+            return ct
+        else:
+            return 0
+
+    if isA(pos, Object):
+        prop['parentOrientation'] = 3
+        obj_dims = (pos.width, pos.length, pos.height)
+        val = lambda self: {
+            'position': pos.relativize(makeOffset(self, obj_dims, makeContactTol(dist, self.contactTolerance), dx, dy, dz)),
+            'parentOrientation': pos.orientation
+        }
+        new = DelayedArgument({axis, "contactTolerance"}, val)
+    elif isA(pos, OrientedPoint):        # TODO too strict?
+        prop['parentOrientation'] = 3
+        val = lambda self: {
+            'position': pos.relativize(makeOffset(self, (0,0,0), 0, dx, dy, dz)),
+            'parentOrientation': pos.orientation
+        }
+        new = DelayedArgument({axis}, val)
+    else:
+        pos = toVector(pos, f'specifier "{syntax} X" with X not a vector')
+        val = lambda self: {'position': pos.offsetRotated(self.orientation, makeOffset(self, (0,0,0), 0, dx, dy, dz))}
+        new = DelayedArgument({axis, 'orientation'}, val)
+    return Specifier(syntax, prop, new)
 
 def Following(field, dist, fromPt=None):
-	"""The :specifier:`following {F} from {X} for {D}` specifier.
+    """The :specifier:`following {F} from {X} for {D}` specifier.
+
+    Specifies :prop:`position` and :prop:`parentOrientation`, with no dependencies.
+
+    Allowed forms::
+
+        following <field> [from <vector>] for <number>
+
+    If the :grammar:`from <vector>` is omitted, the position of ego is used.
+    """
+    if fromPt is None:
+        fromPt = ego()
+    if not isA(field, VectorField):
```

Comment:
For simple cases like this where we are just going to raise an error if there's a type mismatch, it's better to use `toType` instead of `isA`, since the former has fewer false positives (it supports lazy values, deferring type-checking when needed). It's only necessary to use `isA` in cases where we have to immediately know something about the type in order to decide which operations to perform.